src/stochastic_swap.rs

```diff
@@ -0,0 +1,257 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+// Needed to pass shared state between functions
+// closures don't work because of recurssion
+#![allow(clippy::too_many_arguments)]
+
+use std::sync::RwLock;
+
+use hashbrown::HashSet;
+
+use ndarray::prelude::*;
+use numpy::{PyReadonlyArray1, PyReadonlyArray2};
+use rayon::prelude::*;
+
+use pyo3::prelude::*;
+use rand::prelude::*;
+use rand_distr::{Distribution, Normal};
+use rand_pcg::Pcg64Mcg;
+
+use crate::edge_collections::EdgeCollection;
+use crate::nlayout::NLayout;
+
+#[inline]
+fn compute_cost(
+    dist: &ArrayView2<f64>,
+    layout: &NLayout,
+    gates: &[usize],
+    num_gates: usize,
+) -> f64 {
+    (0..num_gates)
+        .map(|kk| {
+            let ii = layout.logic_to_phys[gates[2 * kk]];
+            let jj = layout.logic_to_phys[gates[2 * kk + 1]];
+            dist[[ii, jj]]
+        })
+        .fold(0.0, |a, b| a + b)
+}
+
+/// Computes the symmetric random scaling (perturbation) matrix,
+/// and places the values in the 'scale' array.
+///
+/// Args:
+///     scale (ndarray): An array of doubles where the values are to be stored.
+///     cdist2 (ndarray): Array representing the coupling map distance squared.
+///     rand (double *): Array of rands of length num_qubits*(num_qubits+1)//2.
+///     num_qubits (int): Number of physical qubits.
+#[inline]
+fn compute_random_scaling(
+    scale: &mut Array2<f64>,
+    cdist2: &ArrayView2<f64>,
+    rand: &[f64],
+    num_qubits: usize,
+) {
+    let mut idx: usize = 0;
+    for ii in 0..num_qubits {
+        for jj in 0..ii {
+            scale[[ii, jj]] = rand[idx] * cdist2[[ii, jj]];
+            scale[[jj, ii]] = scale[[ii, jj]];
+            idx += 1
+        }
+    }
+}
+
+fn swap_trial(
+    num_qubits: usize,
+    int_layout: &NLayout,
+    int_qubit_subset: &[usize],
+    gates: &[usize],
+    cdist: ArrayView2<f64>,
+    cdist2: ArrayView2<f64>,
+    edges: &[usize],
+    seed: Option<u64>,
+    trial_num: u64,
+    locked_best_possible: &RwLock<&mut Option<(f64, EdgeCollection, NLayout, usize)>>,
+) -> Option<(f64, EdgeCollection, NLayout, usize)> {
+    {
+        let best_possible = locked_best_possible.read().unwrap();
+        if best_possible.is_some() {
+            return None;
+        }
+    }
+    let mut opt_edges: EdgeCollection = EdgeCollection::new();
```

Comment:
Just because I'm new-ish to Rust: is the type annotation necessary here?  Wouldn't it be inferred from `EdgeCollection::new();` having only one possible type?

Reply:
Yep, it's not necessary here, the compiler can figure out the type implicitly. I think I just did this as a hint to my self as I wrote the `EdgeCollection` struct second and did this module first.